### ISSUE TITLE
Implement isClose / isNear

### ISSUE THREAD

**Issue by dalboris** [vgc/vgc#346]
We need convenient ways to test for almost equality for floating point values and Vec.

We've already hard-coded such a function within another function for unit tests, since the EXPECT_NEAR provided by Google Test was using an absolute tolerance, not a relative tolerance as we needed, see [test_parse.cpp](https://github.com/vgc/vgc/blob/7df33db4bc363333f7abcebc89195e97b7bf1cc3/libs/vgc/core/tests/test_parse.cpp#L49):

```
void readDoubleApproxExpectNear(const std::vector<std::string>& v)
{
    setCLocale();
    for (const std::string& s : v) {
        vgc::core::StringReader in(s);
        double parsed = vgc::core::readDoubleApprox(in);
        double expected = std::stod(s);
        double m = (std::min)(std::abs(parsed), std::abs(expected));
        if (m > 0) {
            EXPECT_LT((parsed-expected)/m, 1e-15) << "Tested string: \"" << s << "\"";
        }
        else {
            EXPECT_EQ(parsed, expected) << "Tested string: \"" << s << "\"";
        }
        // Note: Google Test provides EXPECT_NEAR, but it just computes the
        // absolute difference. For example, 1e-30 and 1e-50 should not be
        // considered near at all, but would pass the following test:
        //   EXPECT_NEAR(1e-30, 1e-50, 1e-15);
        // Conversely, 123456789012345678 and 123456789012345600 should be
        // considered near (15 identical significant digits), but would fail
        // the following test:
        //   EXPECT_NEAR(123456789012345678, 123456789012345600, 1e-15);
    }
}
```

However, the use case I have now if for testing if the size of a widget has changed. In this case, an absolute tolerance is in fact preferable, since a size of `0` is actually typical (the size is zero at initialization), and comparing, say "0 to 0.01" should give the same than comparing, say, "640 to 640.01". The same reasoning applies for inclusion test: testing if the mouse is within the bound of some widget should behave the same next to the "0" bound than next to the "width" bound. These scenarios require an absolute tolerance because we do know the typical magnitudes of these values (a virtual/physical pixel is within, say [-10k, 10k], including hdpi screens and multiple screens setups), and we do know what absolute difference is meaningless and wouldn't produce any visible difference in the rendered output (change a position by 0.2px can be visible due to anti-aliasing, but changes below 0.01px aren't visible to the naked eye).

Therefore, we do need the flexibility to use either absolute and relative comparisons. Python has [math.isclose](https://docs.python.org/3/library/math.html#math.isclose) as built-in, which is quite nice and makes it possible to specify both/either an absolute tolerance and a relative tolerance. Note that [`numpy.isclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html) has a similar API, but with arguably worse default and implementation (see very interesting discussion in https://github.com/numpy/numpy/issues/10161, discussing the flaws in the implementation/defaults).

Unfortunately, in C++, we don't have named arguments so we can't have an API as nice as Python. If we choose to make rel_tol the first optional argument, then it's inconvenient to use an abs_tol with default rel_tol (and vice-versa). So it probably makes sense to have two functions with different names. Since Google Test uses "near" for absolute comparisons, and since Python `math.isclose()` is a relative comparison by default (if you don't provide any optional arguments), I think a good design for VGC is to have both `isClose` for relative comparisons, and `isNear` for absolute comparisons. If you need both a comparison with combined absolute and relative tolerance, then we can have a second optional arguments for `isClose`, like in Python.

**Comment by dalboris** [comment]
An interesting design decision is: should `isClose(inf, inf)` and/or `isNear(inf, inf)` return true or false?

In the case of `isNear(a, b, tol)` (absolute tolerance), when no such helper function is available, people usually write the following one-liner:`if (abs(a-b) < tol) { ... }`. If both `a` and `b` are infinity with the same sign, then their difference is NaN, so abs(a-b) is NaN too, and the comparison returns false (see below).

Returning false is indeed consistent with the fact that (inf - inf) returns NaN instead of zero. The argument would be that some infinities can be greater than other, so by default we can't conclude that two positive infinities are "close/near".

However, `inf == inf` does return true... so it would indeed be surprising if `isClose/Near(inf, inf)` didn't return true too! This is probably why  Python's `math.isclose(inf, inf)` does return true.

However the fact that `inf == inf` returns true is quite inconsistent with the fact that `inf - inf` returns NaN. And since using equality for floating point is generally discouraged, while using difference is okay, it is perhaps better to stay consistent with the behavior of `inf - inf` (i.e., isClose(inf, inf) should return false), rather than being consistent with the behavior of `inf == inf` (i.e., isClose(inf, inf) should return true).

In practice, it should be extremely rare that the function `isClose` or `isNear` is called with two arguments equal to infinity. In fact, I cannot think of a use case. In some algorithms (e.g., finding the min value in a container), you sometime initialize a variable to infinity, but in this case you test for "if smaller than", not "is close to". It seems that anytime you would end up with `isClose(inf, inf)`, is when the infinity value wasn't explicitly initialized to infinity, but rather the result of an overflowing operation. In this case, it does make more sense to return false. This combined with the fact that returning false in those case is a faster implementation, I'm leaning towards this design, that is, do something different than what Python does.

```
#include <iostream>
#include <iomanip>
#include <limits>

int main()
{
    float inf = std::numeric_limits<float>::infinity();
    float ninf = - inf;
    float tol = 0.1;
    
    std::cout << std::abs(inf - inf) << std::endl;
    std::cout << std::abs(ninf - ninf) << std::endl;
    std::cout << std::abs(inf - ninf) << std::endl;
    std::cout << std::abs(ninf - inf) << std::endl << std::endl;
    
    std::cout << std::boolalpha << (std::abs(inf - inf) < tol) << std::endl;
    std::cout << std::boolalpha << (std::abs(ninf - ninf) < tol) << std::endl;
    std::cout << std::boolalpha << (std::abs(inf - ninf) < tol) << std::endl;
    std::cout << std::boolalpha << (std::abs(ninf - inf) < tol) << std::endl << std::endl;
    
    std::cout << std::boolalpha << (inf == inf) << std::endl;
    std::cout << std::boolalpha << (ninf == ninf) << std::endl;
    std::cout << std::boolalpha << (inf == ninf) << std::endl;
    
    return 0;
}
```

Output:

```
nan
nan
inf
inf

false
false
false
false

true
true
false
```

**Comment by dalboris** [comment]
I may have just changed my mind, due to more thinking about why there is such apparent inconsistency between `inf - inf` (returns NaN), and `inf == inf` (returns true).

The thing is, NaN really means "I don't know what the result of this should be". This is why `inf - inf` and `0.0 / 0.0` return NaN: there is not enough info to give a definite answer.

With this philosophy, `inf == inf` should be NaN too: the answer is unknown. However, the boolean type (or integer types, for that matter), do not have a way to represent a NaN value. So some answer *must* be chosen anyway. And it chooses that `inf` is indeed equal to `inf`.

In other words, when comparing `inf` with `inf`, the stance of IEEE 754 is: "They may or may not be equal, we can't know for sure. If we can express the fact that we don't know, they let's do this. Otherwise, if we must decide, let's say that they are indeed equal".

With this interpretation, which actually makes sense, we should have `isClose(inf, inf) == true`, that is, the same behavior as Python.

**Comment by dalboris** [comment]
When implementing isClose/Near for Vec types, one design decision is whether we should use `isClose(a, b)` or `a.isClose(b)`. If we were only using C++, then I would favor `isClose(a, b)`, since this makes the usage consistent with the case when `a` and `b` are fundamental types rather than Vec types, and highlight the symmetric property of the binary operation a little more.

Unfortunately, we also have Python bindings, and there are two annoying things in Python with the "free function" approach:
1. Python does not the concept of ADL (argument-dependent lookup). This means that you would have to do `vgc.core.isClose(a, b)` rather than just `a.isClose(b)`.
2. Python does not (really) support function overloading. It is usually emulated (e.g., by pybind11, or when writing functions directly in Python) at runtime by checking the types and manually dispatching. One can also use tricks such as defining special methods, say `Vec2d.__isClose__`, and have the generic `vgc.core.isClose(a, b)` call `a.__isClose__(b)` if available.

Therefore, in Python, although there are a few workarounds, it generally makes more sense to avoid overloading when possible, and instead use member functions. This makes `a.isClose(b)` a probably more pythonic choice, and convenient for users. Member funcions are also usually more discoverable in the documentation, or via automatic completion.

For consistency between C++ and Python, I'm therefore favoring `a.isClose(b)` both in Python and C++. It's a little less idiomatic C++, but I think a reasonable compromise.

**Comment by dalboris** [comment]
Note: the same reasoning applies for the dot product. Currently, `dot(a, b)` is a free function, and should probably be made a member function `a.dot(b)` instead (like what Eigen does, by the way). 

However, note that numpy does use a lot of free functions, including [`dot(a, b)`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.dot.html?highlight=dot#numpy.dot), but it is important to note that `numpy.ndarray` is one single type, with runtime information about its dimension (= 1 for vectors, 2 for matrices, etc.), shape (= (n, m) for matrices) and dtype (= int32, float64, etc.). This is a different approach than our low-dimensional Vec types which have all different types, and therefore different functions (e.g., 3D cross product not available on Vec2d). This difference may warrant a different approach regarding the use of free functions.

**Comment by dalboris** [comment]
Alright, so handling infinite values becomes a bit messy when generalizing isClose/isNear to points in 2D, 3D, etc. For example, should `(inf, 0)` and `(inf, 42)` be considered close? near?

Let's use the notation `=c=` for "is close" and `=n=` for "is near", assuming (relTol, absTol) = (1e-9, 0.0) for "is close", and absTol = 1e-9 for "is near".

With the current (and Python's) design decision of having `inf =c= inf`, it seems pretty clear that for the 2D case, we should at least have:

```
( inf,  inf) =c= ( inf,  inf)
(-inf,  inf) =c= (-inf,  inf)
( inf, -inf) =c= ( inf, -inf)
(-inf, -inf) =c= (-inf, -inf)
( inf, 42.0) =c= ( inf, 42.0)
```

In other words, if `a == b`, we should have `a =c= b`. The 1D implementation of Python (and our current implementation), is in fact the following (removing error handing of negative tolerances):

```
bool isClose(double a, double b, double relTol, double absTol)
{
    if (a == b) {
        // handle two equal infinities, and maybe speedup exact equality
        return true;
    }
    if (isinf(a) || isinf(b)) {
        // handle two opposite infinities, and finite/infinite comparisons
        return false;
    }
    double diff = fabs(b - a);
    return diff <= fabs(relTol * b) ||
           diff <= fabs(relTol * a) ||
           diff <= absTol;
}
```
However, quid of the following?

```
(inf, 42.0) =c= (inf, 42.0000000001) // => ?
(inf, 42.0) =c= (inf, 43.0)          // => ?
(inf, 42.0) =c= (inf, inf)           // => ?

(inf, 42.0) =n= (inf, 42.0000000001) // => ?
(inf, 42.0) =n= (inf, 43.0)          // => ?
(inf, 42.0) =n= (inf, inf)           // => ?
```

I believe that the "least surprising" behavior which is still mathematically correct and consistent with the 1D behavior of Python should give the following results:

```
---- Behavior A ----

(inf, 42.0) =c= (inf, 42.0)          // => true
(inf, 42.0) =c= (inf, 42.0000000001) // => true
(inf, 42.0) =c= (inf, 43.0)          // => true (!)
(inf, 42.0) =c= (inf, inf)           // => false

(inf, 42.0) =n= (inf, 42.0)          // => true
(inf, 42.0) =n= (inf, 42.0000000001) // => true
(inf, 42.0) =n= (inf, 43.0)          // => false
(inf, 42.0) =n= (inf, inf)           // => false
```

The second line is perhaps the only really surprising. But it comes from the idea that for a sufficiently large finite `x` value, we do have `(x, 42.0) =c= (x, 43.0)`. Indeed, for any `x`, the euclidean distance between those two 2D points is `1.0`, which becomes relatively infinitely smaller than the euclidean length of either `(x, 42.0)` or `(x, 43.0)`, as `x` approaches infinity.

For example, we do have `(1e20, 42.0) =c= (1e20, 43.0)` with the following possible implementation of isClose for finite Vec2d values:

```
bool isClose(const Vec2d& a, const Vec2d& b, double relTol = 1e-9, double absTol = 0.0) const
{
    double diff = (b - a).length();
    return diff <= relTol * b.length() ||
           diff <= relTol * a.length() ||
           diff <= absTol;
}
```

(note: an actual implementation should typically use squared lengths instead for performance)

However, `(inf, 42.0) =n= (inf, 43.0)` should be false, because `(x, 42.0) =n= (x, 43.0)` is false all possible finite values of `x`. For example, `(1e20, 42.0) =n= (1e20, 43.0)` is false with the following possible implementation of isNear for finite Vec2d values:

```
bool isNear(const Vec2d& a, const Vec2d& b, double absTol = 1e-9) const
{
    double diff = (b - a).length();
    return diff <= absTol;
}
```

The important questions are:

1. Do we really want the above behavior A? It might be more confusing/surprising/hard-to-explain than the following simpler behavior B: "isClose and isNear returns false if any of the coordinates is infinity or NaN. In particular, isClose(inf, inf) returns false, despite IEEE 754 specifying that inf == inf.":

```
---- Behavior B ----

(inf, 42.0) =c= (inf, 42.0)          // => false
(inf, 42.0) =c= (inf, 42.0000000001) // => false
(inf, 42.0) =c= (inf, 43.0)          // => false
(inf, 42.0) =c= (inf, inf)           // => false

(inf, 42.0) =n= (inf, 42.0)          // => false
(inf, 42.0) =n= (inf, 42.0000000001) // => false
(inf, 42.0) =n= (inf, 43.0)          // => false
(inf, 42.0) =n= (inf, inf)           // => false
```

2. Is the added value of behavior A worth the confusion, development time, and performance hit, when compared to behavior B?

Indeed, even though the implementation of behavior A is simple enough for the 1D case, it really becomes more complicated in higher dimensions, while behavior B is trivial to implement, and the simpler implementation probably makes the function more likely to be inlined and optimized by compilers.

For example, here is a possible actual implementation of isClose and isNear for behavior A:

```
bool isCloseA(const Vec2d& a, const Vec2d& b, double relTol = 1e-9, double absTol = 0.0) const
{
    if (a == b) {
        return true;
    }
    bool a0inf = isinf(a[0]); bool b0inf = isinf(b[0]);
    bool a1inf = isinf(a[1]); bool b1inf = isinf(b[1]);
    if (a0inf || a1inf || b0inf || b1inf) {
        // If at least one coordinate is infinite, then the number of
        // infinite coordinates is either 1, 2, 3, or 4.
        // 1 => false
        // 3 => false
        // 4 => false because we already know a != b
        // 2 => true iff the two infinite coords have same sign and are
        //       either both x coords, or both y coords.
        if      (a0inf && b0inf && !a1inf && !b1inf) { return a[0] == b[0]; }
        else if (a1inf && b1inf && !a0inf && !b0inf) { return a[1] == b[1]; }
        else { return false; }
    }
    double relTol2 = relTol * relTol;
    double absTol2 = absTol * absTol;
    double diff2 = (b - a).squaredLength();
    return diff2 <= relTol2 * b.squaredLength() ||
           diff2 <= relTol2 * a.squaredLength() ||
           diff2 <= absTol2;
}

bool isNearA(const Vec2d& a, const Vec2d& b, double absTol) const
{
    const Vec2d& a = *this;
    if (a == b) {
        return true;
    }
    bool a0inf = isinf(a[0]); bool b0inf = isinf(b[0]);
    bool a1inf = isinf(a[1]); bool b1inf = isinf(b[1]);
    if (a0inf || a1inf || b0inf || b1inf) {
        if      (a0inf && b0inf && !a1inf && !b1inf) { return a[0] == b[0] && isNear(a[1], b[1], absTol); }
        else if (a1inf && b1inf && !a0inf && !b0inf) { return a[1] == b[1] && isNear(a[0], b[0], absTol); }
        else { return false; }
    }
    if (std::isinf(a[0]) || std::isinf(a[1]) ||
        std::isinf(b[0]) || std::isinf(b[1])) {
        return false;
    }
    double absTol2 = absTol * absTol;
    double diff2 = (b - a).squaredLength();
    return diff2 <= absTol2;
}
```

Compared to a possible actual implementation of behavior B:

```
bool isCloseB(const Vec2d& a, const Vec2d& b, double relTol = 1e-9, double absTol = 0.0) const
{
    double relTol2 = relTol * relTol;
    double absTol2 = absTol * absTol;
    double diff2 = (b - a).squaredLength();
    return diff2 <= relTol2 * b.squaredLength() ||
           diff2 <= relTol2 * a.squaredLength() ||
           diff2 <= absTol2;
}

bool isNearB(const Vec2d& a, const Vec2d& b, double absTol) const
{
    double absTol2 = absTol * absTol;
    double diff2 = (b - a).squaredLength();
    return diff2 <= absTol2;
}
```

The generalization to 3D and 4D of behavior A becomes even more messy. We can define the "infinity-signature" of a floating point as:

```
int infsig(double x) {
    if (x == +inf) { return 1; }
    else if (x == -inf) {return -1; }
    else { return 0;}
}
```

Then, we can define the infinity-signature of a vector as `infsig(v) = (infsig(v[0]), infsig(v[1]), ... )`. 

With this definition, we can implement behavior A, with the following algorithm:
1. Compute the infsigs of both vectors
2. If both are (0, 0, ..., 0), then all values are finite, use normal algorithm (same as behavior B)
3. Otherwise, compare the signatures. If they are different, then return false.
4. Otherwise, for isClose, return true. For isNear, return true is and only if the projection of the two vectors to their finite coordinates are "near", using the normal, finite algorithm of lower dimension.

I'm still a bit undecided of whether go that route, or simply use behavior B (and break consistency with Python `math.isClose`)

**Comment by dalboris** [comment]
Another interesting thread I've just read: https://forums.swift.org/t/se-0259-approximate-equality-for-floating-point/23627/73 . It is a recent discussion (Apr-May 2019) about a potential (not yet approved) addition of `isAlmostEqual()` to the standard library of Swift.

It doesn't discuss our dilemna regarding the handling of infinities, but it discusses a lot about the behavior around zero, and comforts me in my decision of having both `isNear()` and `isClose()` methods, where isClose has a default value for both relTol and absTol, while isNear doesn't. Our documentation makes it clear of the difference, and guides the user towards the appropriate method to use. Most points discussed in the thread are quite nicely resolved by VGC's implementation.

**Comment by dalboris** [comment]
Regarding how to handle infinities, it appears clear that the complexity of implementing behavior A stems from the fact that `inf - inf = NaN` rather than zero, despite having `inf == inf`, and therefore the desired behavior that `isClose(inf, inf) == isNear(inf, inf) == true`.

We could simplify the implementation by using a custom `infdiff(a, b)` function which always returns zero if `a == b`, including when `a` and `b` are infinities of the same sign. For example:

```
/// Computes the difference `a - b`, but where two infinities of the same sign
/// are considered exactly equal, so their difference is zero.
///
/// ```cpp
/// double inf = std::numeric_limits<double>::infinity();
/// std::cout << inf - inf << std::endl;         // "nan"
/// std::cout << infdiff(inf, inf) << std::endl; // "0"
/// }
/// ```
///
double infdiff(double a, double b)
{
    return (a == b) ? 0.0 : (a - b);
}
```

In fact, there is the function [`relative_difference(a, b)`](https://www.boost.org/doc/libs/1_67_0/libs/math/doc/html/math_toolkit/float_comparison.html) in Boost which does return 0 if both a and b are infinity of the same sign. In our case, we do want to compute an absolute difference, but it's an interesting comparison.

With the above `infdiff` function, it seems that the implementation of behavior A should be more straightforward, and still reasonably fast. For example:

```
Vec2d infdiff(const Vec2d& a, const Vec2d& b)
{
    return Vec2d(infdiff(a[0], b[0]), infdiff(a[1], b[1]));
}

bool isCloseA(const Vec2d& a, const Vec2d& b, double relTol = 1e-9, double absTol = 0.0)
{
    double relTol2 = relTol * relTol;
    double absTol2 = absTol * absTol;
    double diff2 = infdiff(a, b).squaredLength();
    return diff2 <= relTol2 * b.squaredLength() ||
           diff2 <= relTol2 * a.squaredLength() ||
           diff2 <= absTol2;
}

bool isNearA(const Vec2d& a, const Vec2d& b, double absTol)
{
    double absTol2 = absTol * absTol;
    double diff2 = infdiff(a, b).squaredLength();
    return diff2 <= absTol2;
}
```

I should still do a bit more analysis, but this seems like a good direction to pursue.

**Comment by dalboris** [comment]
Alright, so in the end, I did implement behavior A, using the `infdiff` method mentioned in the previous post, which works like a charm. Not only does it make implementing behavior A for all dimensions trivial, but in fact it results in one less comparison than our previous implementation, even in the 1D case. This means that I expect the VGC implementation of the 1D case to actually be slighlty faster than the Python implementation (at least when the values are not equal), despite having exactly the same behavior. Also, we never call `std::isinf()`, but instead only test for equality with positive infinite: `diff == std::numeric_limits<T>::infinity()`.

The implementation in the above post did have one bug though: `isClose(inf, -inf)` returned true, because `diff2` was +infinite, but `relTol2 * a.squaredLength()` was infinite too, so `diff2 <= relTol2 * a.squaredLength()` was true. The trick to solve this is simply to check whether `diff2` is +infinite before further processing, which correctly handles all possible cases both for the 1D case and n-D cases. I have added a lot of tests to ensure that it works ok.

The final implementation is the following (I removed template declarations for readability):

```
T infdiff(T a, T b)
{
    return (a == b) ? 0 : (a - b);
}

bool isClose(T a, T b, T relTol, T absTol)
{
    T diff = std::fabs(infdiff(a, b));
    if (diff == std::numeric_limits<T>::infinity()) {
        return false; // opposite infinities or finite/infinite mismatch
    }
    else {
        return diff <= std::fabs(relTol * b) ||
               diff <= std::fabs(relTol * a) ||
               diff <= absTol;
    }
}

bool isNear(T a, T b, T absTol)
{
    T diff = std::fabs(infdiff(a, b));
    if (diff == std::numeric_limits<T>::infinity()) {
        return false; // opposite infinities or finite/infinite mismatch
    }
    else {
        return diff <= absTol;
    }
}

Vec2d infdiff_(const Vec2d& b) const
{
    const Vec2d& a = *this;
    return Vec2d(internal::infdiff(a[0], b[0]),
                 internal::infdiff(a[1], b[1]));
}

bool isClose(const Vec2d& b, double relTol = 1e-9, double absTol = 0.0) const
{
    const Vec2d& a = *this;
    double diff2 = a.infdiff_(b).squaredLength();
    if (diff2 == std::numeric_limits<double>::infinity()) {
        return false; // opposite infinities or finite/infinite mismatch
    }
    else {
        double relTol2 = relTol * relTol;
        double absTol2 = absTol * absTol;
        return diff2 <= relTol2 * b.squaredLength() ||
               diff2 <= relTol2 * a.squaredLength() ||
               diff2 <= absTol2;
    }
}

bool isNear(const Vec2d& b, double absTol) const
{
    const Vec2d& a = *this;
    double diff2 = a.infdiff_(b).squaredLength();
    if (diff2 == std::numeric_limits<double>::infinity()) {
        return false; // opposite infinities or finite/infinite mismatch
    }
    else {
        double absTol2 = absTol * absTol;
        return diff2 <= absTol2;
    }
}
```

For the n-D case, I also added per-coordinate versions of `isClose()` and `isNear()`, which are called `allClose()` and `allNear()` (the naming refers to `std::all_of()`). In other words, `a.allClose(b)` (resp. `a.allNear(b)`) is true if and only if `isClose(a[i], b[i])` (resp. `isNear(a[i], b[i])`) is true for all coordinates `i`. All four proximity tests have different behaviors and are useful in different scenarios.
